genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/JobDirectoryLimitsServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+import com.netflix.genie.common.internal.services.JobDirectoryManifestCreatorService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.unit.DataSize;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * Implementation of {@link JobDirectoryLimitsService} that periodically checks on the size and number of files
+ * using the manifest creator, rather than looking at the actual files.
+ * This implementation is not thread safe and does not check for double start / double stop.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+class JobDirectoryLimitsServiceImpl implements JobDirectoryLimitsService {
+    private final KillService killService;
+    private final JobDirectoryManifestCreatorService manifestCreatorService;
+    private final TaskScheduler taskScheduler;
+    private final JobDirectoryLimitsServiceProperties properties;
+    private ScheduledFuture<?> scheduledCheck;
+
+    JobDirectoryLimitsServiceImpl(
+        final KillService killService,
+        final JobDirectoryManifestCreatorService manifestCreatorService,
+        final TaskScheduler taskScheduler,
+        final AgentProperties agentProperties) {
+        this.killService = killService;
+        this.manifestCreatorService = manifestCreatorService;
+        this.taskScheduler = taskScheduler;
+        this.properties = agentProperties.getJobDirectoryLimitsService();
+    }
+
+    @Override
+    public void start(final Path jobDirectory) {
+        this.scheduledCheck = this.taskScheduler.scheduleAtFixedRate(
+            () -> this.checkFilesSize(jobDirectory),
+            this.properties.getCheckInterval()
+        );
+
+    }
+
+    @Override
+    public void stop() {
+        if (this.scheduledCheck != null) {
+            this.scheduledCheck.cancel(true);
+        }
+    }
+
+    private void checkFilesSize(final Path jobDirectory) {
+        final DirectoryManifest manifest;
+        try {
+            manifest = this.manifestCreatorService.getDirectoryManifest(jobDirectory);
+        } catch (IOException e) {
+            log.warn("Failed to obtain manifest: {}" + e.getMessage());
+            return;
+        }
+
+        final int files = manifest.getNumFiles();
+        final int maxFiles = this.properties.getMaxFiles();
+        if (files > maxFiles) {
+            log.warn("Limit exceeded, too many files: {}/{}", files, maxFiles);
+            this.killService.kill(KillService.KillSource.FILES_LIMIT);
+            return;
+        }
+
+        final DataSize totalSize = DataSize.ofBytes(manifest.getTotalSizeOfFiles());
+        final DataSize maxTotalSize = this.properties.getMaxTotalSize();
+        if (totalSize.toBytes() > maxTotalSize.toBytes()) {
+            log.warn("Limit exceeded, job directory too large: {}/{}", totalSize, maxTotalSize);
```

Comment:
Since you're about to kill it I'd think we should make these `error` level messages

Reply:
Done